src/goto-instrument/cover.cpp

```diff
@@ -11,1687 +11,196 @@ Date: May 2016
-  bool ignore_trivial)
+  message_handlert &message_handler)
 {
-  // exclude trivial coverage goals of a goto program
-  if(ignore_trivial && program_is_trivial(goto_program))
-    return;
+  goal_filterst goal_filters;
+  goal_filters.add(util_make_unique<internal_goals_filtert>(message_handler));
 
-  // ignore if built-in library
-  if(!goto_program.instructions.empty() &&
-     goto_program.instructions.front().source_location.is_built_in())
-    return;
+  cover_instrumenterst instrumenters;
+  instrumenters.add_from_criterion(criterion, symbol_table, goal_filters);
 
-  const namespacet ns(symbol_table);
-  basic_blockst basic_blocks(goto_program);
-  basic_blocks.select_unique_java_bytecode_indices(
-    goto_program, message_handler);
-  basic_blocks.report_block_anomalies(goto_program, message_handler);
-
-  const irep_idt coverage_criterion=as_string(criterion);
-  const irep_idt property_class="coverage";
+  instrument_cover_goals(
+    goto_program, instrumenters, message_handler);
+}
 
-  Forall_goto_program_instructions(i_it, goto_program)
+/// Create and add an instrumenter based on the given criterion
+/// \param criterion: the coverage criterion
+/// \param symbol_table: the symbol table
+/// \param goal_filters: goal filters to discard certain goals
+void cover_instrumenterst::add_from_criterion(
+  coverage_criteriont criterion,
+  const symbol_tablet &symbol_table,
+  const goal_filterst &goal_filters)
+{
+  switch(criterion)
+  {
+  case coverage_criteriont::LOCATION:
+    instrumenters.push_back(util_make_unique<cover_location_instrumentert>(
+      symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::BRANCH:
+    instrumenters.push_back(
+      util_make_unique<cover_branch_instrumentert>(symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::DECISION:
+    instrumenters.push_back(util_make_unique<cover_decision_instrumentert>(
+      symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::CONDITION:
+    instrumenters.push_back(util_make_unique<cover_condition_instrumentert>(
+      symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::PATH:
+    instrumenters.push_back(
+      util_make_unique<cover_path_instrumentert>(symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::MCDC:
+    instrumenters.push_back(
+      util_make_unique<cover_mcdc_instrumentert>(symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::ASSERTION:
+    instrumenters.push_back(util_make_unique<cover_assertion_instrumentert>(
+      symbol_table, goal_filters));
+    break;
+  case coverage_criteriont::COVER:
+    instrumenters.push_back(
+      util_make_unique<cover_cover_instrumentert>(symbol_table, goal_filters));
+  }
+}
+
+/// Parses a coverage criterion
+/// \param criterion_string: a string
+/// \return a coverage criterion or throws an exception
+coverage_criteriont
+parse_coverage_criterion(const std::string &criterion_string)
+{
```

Comment:
Command line parsing is a bit involved here. Repeating the code for each binary cannot be the solution. I've provided utility functions here that should be used by the binary to avoid duplication.